### PR TITLE
KEYCLOAK-12273 Operator documentation

### PR DESCRIPTION
**Pull request by slaskawi**
[KEYCLOAK-12273](https://issues.jboss.org/browse/KEYCLOAK-12273)

This is an initial version of Keycloak Operator documentation. @abstractj @stianst @hmlnarik @andymunro Please have a look at it when you find a moment.

Still to be done:
* [x] Monitoring Stack paragraph
* [x] Reference new documentation from the old one

### CODE REVIEW

File: server_installation/topics/operator/monitoring-stack.adoc [hunk 1]
```diff
@@ -0,0 +1,31 @@
+
+=== Monitoring Stack
```
Comment (slaskawi) [inline server_installation/topics/operator/monitoring-stack.adoc:2]: Hey @pb82 ! If I could ask you to review this part.

File: server_installation/topics/operator.adoc [hunk 2]
```diff
@@ -0,0 +1,14 @@
+
+[[_operator]]
+== {project_operator}
+
+{project_operator} has been designed to automate {project_name} administration in Kubernetes/OpenShift cluster.
+It allows you to:
+
+* Install Keycloak to a namespace
```
Comment (iankko) [inline server_installation/topics/operator.adoc:8]: @slaskawi If examples of steps, that need to be done to achieve this, are provided later in the docs for each of these subcases, would it be possible to make each of these points a link to the section, providing those steps?

File: server_installation/topics/operator/external-database.adoc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    POSTGRES_PASSWORD: <Database Password>
+    # Required for AWS Backup functionality
+    POSTGRES_SUPERUSER: true
+    POSTGRES_USERNAME: <Database Username>
+ type: Opaque
+```
+
+There are two properties specifically responsible for specifying external database hostname or IP address and its port:
+
+* POSTGRES_EXTERNAL_ADDRESS - an IP address (or a Hostname) or the external database. This address needs to be resolvable from Kubernetes cluster perspective.
```
Comment (iankko) [inline server_installation/topics/operator/external-database.adoc]: 1) Typo - ```s/or the external database/of the external database/```
2) Rephrase `This address needs to be resolvable from Kubernetes cluster perspective` to `This address needs to be resolvable from the perspective of a Kubernetes cluster.`

File: server_installation/topics/operator/external-database.adoc [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  namespace: keycloak
+spec:
+  externalDatabase:
+    enabled: true
+  instances: 1
+```
+
+==== Implementation details
+
+{project_operator} allows you to use an external Postgresql database by modifying `keycloak-postgresql` Endpoints object. Typically, this object is maintained by Kubernetes, however it might be used to connect to an external service. See https://docs.openshift.com/container-platform/3.11/dev_guide/integrating_external_services.html[OpenShift manual] for more details.
```
Comment (iankko) [inline server_installation/topics/operator/external-database.adoc]: Is this `keycloak-postgresql` endpoint created dynamically based on the value of the `POSTGRES_HOST` above? Or is it `keycloak-postgresql` always?

If the former, maybe rephrase to:
`{project_operator} allows you to use an external PostgreSQL database by modifying the Endpoints object, with the name matching the value of the POSTGRES_HOST variable (e.g. 'keycloak-postgresql')`.

If the latter, it should be highlighted the `keycloak-postgresql` name of the Endpoint, is something else than the value of the `POSTGRES_HOST` variable value (IOW the Endpoint is always named that way) to prevent confusion.

File: server_installation/topics/operator/installation.adoc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+* Running manually
+
+==== Using Operator Lifecycle Manager
+
+In case of a newly installed OpenShift 4.x cluster, {project_operator} should already be available in
+OperatorHub tab of the OpenShift administration panel (as shown in the image below).
+
+image:{project_images}/operator-openshift-operatorhub.png[]
+
+If {project_operator} is not available, please navigate to https://operatorhub.io/operator/keycloak-operator, click install button and follow the instructions on screen.
```
Comment (iankko) [inline server_installation/topics/operator/installation.adoc]: Maybe "click `Install` button"? (if "Install" is the name of the button)

File: server_installation/topics/operator/installation.adoc [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+image:{project_images}/operator-operatorhub-install.png[]
+
+TIP: For Kubernetes installation, please follow link:https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/install/install.md[Operator Lifecycle Manager Installation Guide]
+
+Once {project_operator} is available in your cluster, click `install` button and follow the instruction on
```
Comment (iankko) [inline server_installation/topics/operator/installation.adoc]: Maybe `Install` ?

File: server_installation/topics/operator/installation.adoc [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+* Validate if {project_operator} is running
+
+```
+kubectl get deployment keycloak-operator
+NAME                READY   UP-TO-DATE   AVAILABLE   AGE
+keycloak-operator   1/1     1            1           41s
+```
+
+TIP: `kubectl apply` commands can be replaced by `oc create`.
```
Comment (iankko) [inline server_installation/topics/operator/installation.adoc]: Maybe "s/TIP: `kubectl apply` commands can be replaced by `oc create`./TIP: When deploying {project_operator} to an OpenShift cluster, `kubectl apply` commands can be replaced with their `oc create` counterparts./" ?

Also, if using ```bash anotation for one code snippet / block, could it be applied to every code snippet / code block in this PR? (so the look is uniform)

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 8]
```diff
@@ -0,0 +1,106 @@
+
+=== KeycloakBackup Custom Resource
+
+{project_operator} provides automatic backups with manual restore in 3 modes:
+
+* One time backups to a local Persistent Volume.
+* One time backups to Amazon S3 storage.
+* Periodic backups to Amazon S3 storage.
+
+The Operator uses `KeycloakBackup` Custom Resource to trigger a backup Job (or a `CronJob` in case of Periodic Backups) and reports back its status. The CR has the following structure:
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: Maybe replace `Custom Resource` with `Custom Resource (CR)` so later in the yaml example below it's clear `CR` refers to custom resource?

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+.`KeycloakBackup` Custom Resource
+```yaml
+apiVersion: keycloak.org/v1alpha1
+kind: KeycloakBackup
+metadata:
+  name: <CR Name>
+spec:
+  aws:
+    # Optional - used only for Periodic Backups
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: Maybe `# Optional - used only for Periodic Backups. Follows usual crond syntax (e.g. use "0 1 * * *") to perform the backup every day at 1 AM` ? If the expected syntax is different, can you provide an example value?

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    schedule: <Cron Job Schedule>
+    # Requred - used to obtain credentials for storing backup in S3
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: Maybe `s/Requred/Required - the name of the secret containing the credentials to access the S3 storage in order to perform the backup`  or `Required - the name of the secret containing the credentials to access the S3 storage`

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+metadata:
+  name: <Secret Name>
+type: Opaque
+stringData:
+  AWS_S3_BUCKET_NAME: <S3 Bucket Name>
+  AWS_ACCESS_KEY_ID: <AWS Access Key ID>
+  AWS_SECRET_ACCESS_KEY: <AWS Secret Key>
+```
+
+The above secret name needs to match the one referred in the `KeycloakBackup` Custom Resource.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: Mark this sentence with `IMPORTANT: ` annotation / Asciidoc elem, so it's not overlooked.

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+Once the `KeycloakBackup` Custom Resource is created, {project_operator} will create a corresponding `Job`
+responsible for backing up Postgresql database. The backup status will be reported in the status field.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: Maybe "will create a corresponding `Job` to back up the PostgreSQL database." ?
Maybe "The status of the backup is reported in the `Status` field."

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  Secondary Resources:
+    Job:
+      example-keycloakbackup
+    Persistent Volume Claim:
+      keycloak-backup-example-keycloakbackup
+```
+
+==== Backups to AWS S3
+
+In order to create Backups uploaded to S3 storage, you need to create a `KeycloakBackup` Custom Resource with `aws` sub-properties. The `credentialsSecretName` field is required and needs to contain a valid reference to a `Secret` containing AWS S3 credentials. If the `schedule` contains valid `CronJob` schedule definition, the Operator will backup your data periodically.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: Again, please annotate the second sentence (... The `credentialsSecretname` field .. AWS S3 credentials.` ) with the `IMPORTANT:` tag (and maybe even dedicate an own paragraph for it), so it's not overlooked.

File: server_installation/topics/operator/keycloak-backup-cr.adoc [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+kind: KeycloakBackup
+metadata:
+  name: <CR Name>
+```
+
+{project_operator} will create a new `PersistentVolumeClaim` with the following naming scheme:
+
+  keycloak-backup-<Custom Resource Name>
+
+It is a good practice to create a corresponding `PersistentVolume` for the backups upfront and use `claimRef` to reserve it only for `PersistentVolumeClaim` created by the Keycloak Operator (see https://docs.okd.io/3.6/dev_guide/persistent_volumes.html#persistent-volumes-volumes-and-claim-prebinding[OKD manual for more details]).
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-backup-cr.adoc]: `for the upcoming backups`? `for the backups to be performed in the future`?

Or "It is a good practice to create a corresponding `PersistentVolume`, dedicated for the backups beforehand and use ..."?

(not sure I got the previous sentence correctly - is it trying to say, the persistent volumes for upcoming backups should be reserved / created beforehand and labeled with the `claimRef` label, so they can be used solely by the Keycloak operator?)

File: server_installation/topics/operator/keycloak-client-cr.adoc [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+{project_operator} allows application developers to represent Keycloak Clients as Custom Resources:
+
+.`KeycloakClient` Custom Resource
+```yaml
+apiVersion: keycloak.org/v1alpha1
+kind: KeycloakClient
+metadata:
+  name: <Keycloak Client name>
+  labels:
+    app: sso
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-client-cr.adoc:13]: Is this `app: sso` label used for something later? (e.g. for a search how many Keycloak clients exist in the corresponding namespace)

IOW does it need always to be set to `app: sso`, since its used for something later, the proper function of the operator relies on, or it can be set to arbitrary value? If it should be always set to `app: sso` it should be mentioned IMHO somewhere here.

File: server_installation/topics/operator/keycloak-client-cr.adoc [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    #id: 123
+    clientId: client-secret
+    secret: client-secret
+    # ...
+    # other properties of Keycloak Client
+```
+
+TIP: Note, that `realmSelector` needs to match labels of an existing `KeycloakRealm` Custom Resource.
+
+{project_operator} synchronizes all the changes made to the Custom Resource with a running {project_name} instance. No manual changes via Keycloak Admin Console are allowed.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-client-cr.adoc]: Add a `NOTE:` annotation here, `NOTE: {project operator} .... No manual changes ... are allowed.`

Or maybe make it even `IMPORTANT:` if manual attempt is prohibited / might lead to error.

File: server_installation/topics/operator/keycloak-client-cr.adoc [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  Message:
+  Phase:    reconciling
+  Ready:    true
+  Secondary Resources:
+    Secret:
+      keycloak-client-secret-client-secret
+Events:  <none>
+```
+
+Once a Client is created, {project_operator} creates a Secret with `Client ID` as well as the client's secret using the following naming pattern: `keycloak-client-secret-<Custom Resource name>`. Here's an example:
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-client-cr.adoc]: Maybe replace "{project_operator} creates a Secret with `Client ID` as well as the client's secret ..." with
"{project_operator} creates a secret, containing both the `Client ID` and client secret using the ..."

(the way it's written now, got me to understand, they are two secrets one with `Client ID` and second with `Client secret`created. The example explained it though)

File: server_installation/topics/operator/keycloak-cr.adoc [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  name: example-keycloak
+  labels:
+    app: sso
+spec:
+  instances: 1
+  externalAccess:
+    enabled: True
+```
+
+The `Spec` contains three objects (two of them are listed the above and the external database has been covered in the other paragraphs of this manual):
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-cr.adoc]: What are those tree objects? `instances`, `externalAccess` and `enabled` ? Not getting, how / why it refers to something listed above - can it be more specific?

File: server_installation/topics/operator/keycloak-cr.adoc [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+* `instances` - controls the number of instances running in HA mode
+* `externalAccess` - if the `enabled` flag is set to `True`, {project_operator} will create a Route or an Ingress (depending on available types in the cluster) for {project_name} cluster.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-cr.adoc]: Maybe "if the `enabled` flag is set to `True`, depending on the types available in the cluster {project_operator} will create a Route or an Ingress for {project_name} cluster."?

File: server_installation/topics/operator/keycloak-realm-cr.adoc [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    id: "basic"
+    realm: "basic"
+    enabled: True
+    displayName: "Basic Realm"
+  instanceSelector:
+    matchLabels:
+      app: sso
+```
+
+TIP: Note, that `instanceSelector` needs to match labels of an existing `Keycloak` Custom Resource.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-realm-cr.adoc]: Turn this into `IMPORTANT:`

File: server_installation/topics/operator/keycloak-realm-cr.adoc [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+{project_operator} synchronizes all the changes made to the Custom Resource with a running {project_name} instance. No manual changes via {project_name} Admin Console are allowed.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-realm-cr.adoc]: This two sentences should be annotated with `IMPORTANT:` too.

File: server_installation/topics/operator/keycloak-user-cr.adoc [hunk 22]
```diff
@@ -0,0 +1,76 @@
+
+=== KeycloakUser Custom Resource
+
+{project_operator} allows application developers to represent {project_name} Users as Custom Resources:
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-user-cr.adoc]: Why these sentences (also for Keycloak realms and clients above) need to be written in so a complex way?

Why not just `{project_name} Users can be represented as Custom Resources in {project_operator}.`?

Or `Use Custom Resource in {project_operator} to represent {project_name} Users`

File: server_installation/topics/operator/keycloak-user-cr.adoc [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+      account:
+        - "manage-account"
+      realm-management:
+        - "manage-users"
+  realmSelector:
+    matchLabels:
+      app: sso
+```
+
+TIP: Note, that `realmSelector` needs to match labels of an existing `KeycloakRealm` Custom Resource.
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-user-cr.adoc]: Change this three lines into `IMPORTANT: ` annotation, so they aren't overlooked.

File: server_installation/topics/operator/keycloak-user-cr.adoc [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+    Last Name:       Doe
+    Username:        realm_user
+Status:
+  Message:
+  Phase:    reconciled
+Events:     <none>
+```
+
+Once a User is created, {project_operator} creates a Secret with the user credentials using the
+following naming pattern: `credential-<realm name>-<user name>-<namespace>`. Here's an example:
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-user-cr.adoc:66]: Please perform the same change like in the `KeycloakClient Secret` example, couple of lines / paragraphs back.

File: server_installation/topics/operator/monitoring-stack.adoc [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+{project_operator} integrates with {application_monitoring_operator} providing Graphana Dashboard and Prometheus Alerts. The integration is seamless and doesn't require any action from the user. However, {application_monitoring_operator} needs to be properly installed in the cluster.
+
+==== Installing {application_monitoring_operator}
+
+Prior to running {project_operator} it is recommended to install {application_monitoring_operator} by following the link:{application_monitoring_operator_installation_link}[documentation]. After {application_monitoring_operator} is up and running, you need to annotate a namespace used for {project_operator} installation. Here's an example:
```
Comment (iankko) [inline server_installation/topics/operator/monitoring-stack.adoc]: `you need to annotate the namespace used for ...`?

File: server_installation/topics/operator/monitoring-stack.adoc [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+
+.Annotating namespace for monitoring
+```bash
+oc label namespace <namespace> monitoring-key=middleware
+```
+
+Once the installation is complete and {application_monitoring_operator}, you can search for Prometheus and Graphana route in `application-monitoring` namespace:
```
Comment (iankko) [inline server_installation/topics/operator/monitoring-stack.adoc]: s/`Once the installation is complete and {application_monitoring_operator},`/`Once the {application_monitoring_operator} is installed and running, also the corresponding namespace is annotated for monitoring, you can search .../` ?

File: server_installation/topics/operator.adoc (hunk 2 above)
Comment (slaskawi) [reply server_installation/topics/operator.adoc:8]: I don't think so.... This is more a list of features implemented. Actually, it's the same list as here: https://operatorhub.io/operator/keycloak-operator

File: server_installation/topics/operator/external-database.adoc (hunk 3 above)
Comment (slaskawi) [reply server_installation/topics/operator/external-database.adoc]: Fixed.

File: server_installation/topics/operator/external-database.adoc (hunk 4 above)
Comment (slaskawi) [reply server_installation/topics/operator/external-database.adoc]: No, it's generated based on Keycloak CR name as follows: `<Keycloak CR Name>-postgresql`. I think I get your point. Let me rephrase this a bit.

The `POSTGRES_HOST` is not strictly required by the Operator itself but rather by Integreately Backup Container, see https://github.com/integr8ly/backup-container-image/blob/master/image/tools/lib/component/postgres.sh#L16

File: server_installation/topics/operator/installation.adoc (hunk 5 above)
Comment (slaskawi) [reply server_installation/topics/operator/installation.adoc]: Fixed,

File: server_installation/topics/operator/installation.adoc (hunk 6 above)
Comment (slaskawi) [reply server_installation/topics/operator/installation.adoc]: Fixed

File: server_installation/topics/operator/installation.adoc (hunk 7 above)
Comment (slaskawi) [reply server_installation/topics/operator/installation.adoc]: Fxied. `bash` annotation applied everywhere.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 8 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Good idea. Fixed.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 9 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Fxied.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 10 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 11 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 12 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 13 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-backup-cr.adoc (hunk 14 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-backup-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-client-cr.adoc (hunk 15 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-client-cr.adoc:13]: It might be used in the future I believe. At the moment, we apply those labels for two reasons:
- If `app` label is there, OpenShift shows them a bit nicer in the UI (e.g. you have a Route available in the StatefulSet deployment since it knows it's the same application).
- `Keycloak` and `KeycloakRealm` labels are used to find proper installation and realm to inject the client to. At the moment, there's nothing looking for existing Clients but I guess it's very likely we'll have something in the future.

File: server_installation/topics/operator/keycloak-client-cr.adoc (hunk 16 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-client-cr.adoc]: I guess `NOTE` is fine.

File: server_installation/topics/operator/keycloak-client-cr.adoc (hunk 17 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-client-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-cr.adoc (hunk 18 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-cr.adoc]: Changed to "properties". Those are CR properties, more technically - Go structs.

File: server_installation/topics/operator/keycloak-cr.adoc (hunk 19 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-cr.adoc]: Fixed

File: server_installation/topics/operator/keycloak-realm-cr.adoc (hunk 20 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-realm-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-realm-cr.adoc (hunk 21 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-realm-cr.adoc]: I added a `NOTE`

File: server_installation/topics/operator/keycloak-user-cr.adoc (hunk 22 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-user-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-user-cr.adoc (hunk 23 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-user-cr.adoc]: Fixed.

File: server_installation/topics/operator/keycloak-user-cr.adoc (hunk 24 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-user-cr.adoc:66]: Fixed.

File: server_installation/topics/operator/monitoring-stack.adoc (hunk 25 above)
Comment (slaskawi) [reply server_installation/topics/operator/monitoring-stack.adoc]: Fixed.

File: server_installation/topics/operator/monitoring-stack.adoc (hunk 26 above)
Comment (slaskawi) [reply server_installation/topics/operator/monitoring-stack.adoc]: Fixed.

**Comment by slaskawi** [comment]
@andymunro I inspected the build logs and it contains a massive amount of false-positives. All redirections to `issues.redhat.com` should probably be fixed. Also there's a false-positive about an anchor I used (which works fine). 

@stianst I guess we need find some way to work this through.

**Comment by slaskawi** [comment]
@iankko Thanks for a detailed review. I addressed all the comments.

File: server_installation/topics/operator/installation.adoc [hunk 27]
```diff
@@ -0,0 +1,68 @@
+
+=== Installing {project_operator} into OpenShift cluster
+
+There are ways of running {project_operator} in Kubernetes/OpenShift cluster:
```
Comment (iankko) [inline server_installation/topics/operator/installation.adoc]: @slaskawi Maybe missing `two` above? Like `There are two ways of running..`

File: server_installation/topics/operator/installation.adoc [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+
+* Validate if {project_operator} is running
+
+```bash
+kubectl get deployment keycloak-operator
+NAME                READY   UP-TO-DATE   AVAILABLE   AGE
+keycloak-operator   1/1     1            1           41s
+```
+
+TIP: TIP: When deploying {project_operator} to an OpenShift cluster, `kubectl apply` commands can be replaced with their `oc create` counterparts.
```
Comment (iankko) [inline server_installation/topics/operator/installation.adoc]: `TIP:` is now 2-times here.

**Review by iankko — DISMISSED** [review]
Just two small issues / nitpicks yet. But LGTM as a whole.

File: server_installation/topics/operator/installation.adoc (hunk 27 above)
Comment (slaskawi) [reply server_installation/topics/operator/installation.adoc]: Fixed

File: server_installation/topics/operator/installation.adoc (hunk 28 above)
Comment (slaskawi) [reply server_installation/topics/operator/installation.adoc]: Fixed

**Comment by slaskawi** [comment]
Thanks @iankko ! Ready for the next round.

File: server_installation/topics/operator/external-database.adoc [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+
+.`keycloak-db-secret` Secret
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+    name: keycloak-db-secret
+    namespace: keycloak
+stringData:
+    POSTGRES_DATABASE: <Database Name>
```
Comment (fidgi) [inline server_installation/topics/operator/external-database.adoc:14]: Why not using the same DB_ prefixed variables as recommended in https://github.com/keycloak/keycloak-containers/tree/master/server ? Having to configure a different set of variables for the DB could be error prone.

File: server_installation/topics/operator/keycloak-cr.adoc [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+  name: example-keycloak
+  labels:
+    app: sso
+spec:
+  instances: 1
+  externalAccess:
+    enabled: True
+```
+
+The `Spec` contains three properties (two of them are listed the above and the external database has been covered in the other paragraphs of this manual):
```
Comment (iankko) [inline server_installation/topics/operator/keycloak-cr.adoc]: @slaskawi `s/two of them are listed the above/two of them are listed above/g`

**Review by iankko — CHANGES_REQUESTED** [review]
Just one -- drop unneeded `the` nitpick. The rest LGTM.

File: server_installation/topics/operator/external-database.adoc (hunk 29 above)
Comment (iankko) [reply server_installation/topics/operator/external-database.adoc:14]: @fidgi The proposal to use [DB_ prefixed variables](https://github.com/keycloak/keycloak-containers/tree/master/server#environment-variables) LGTM. Could you file a new Keycloak Operator JIRA RFE to have this implemented in the future?

(the current documentation is describing the current variable names. From a brief look at the code, looks besides the variable names change in the documentation, also some variable names change would need to be done in the operator code itself, e.g. in [recognized constants](https://github.com/keycloak/keycloak-operator/blob/master/pkg/model/constants.go#L29). Since docs change itself wouldn't be sufficient, this might be a candidate for a future RFE / change)

@slaskawi WDYT?

Comment (slaskawi) [reply server_installation/topics/operator/external-database.adoc:14]: Thanks for the suggestion @fidgi @iankko !

We already have a change request around this in [KEYCLOAK-12040](https://issues.redhat.com/browse/KEYCLOAK-12040). Please follow up using that JIRA.

And of course, @iankko is correct here - the documentation describes the current approach. Supporting other databases is a feature request.

File: server_installation/topics/operator/keycloak-cr.adoc (hunk 30 above)
Comment (slaskawi) [reply server_installation/topics/operator/keycloak-cr.adoc]: Fixed.

**Comment by slaskawi** [comment]
@iankko @stianst @andymunro Rebased and ready to be reviewed.

**Comment by stianst** [comment]
This is not a high priority right now and should be put on hold for now.

**Comment by slaskawi** [comment]
After a short discussion with @andymunro we decided to close this Pull Request and @andymunro will open another one with necessary modifications.
